pallets/anonymity-mining/src/lib.rs

```diff
@@ -192,6 +305,171 @@ pub mod pallet {
+				src_chain_id,
+				merkle_root,
+				latest_leaf_index,
+				src_resource_id,
+			)?;
+			Self::deposit_event(Event::AnchorEdgeUpdated);
+		} else {
+			T::VAnchor::add_edge(
+				tree_id,
+				src_chain_id,
+				merkle_root,
+				latest_leaf_index,
+				src_resource_id,
+			)?;
+			Self::deposit_event(Event::AnchorEdgeAdded);
+		}
+		Ok(().into())
+	}
+
+	fn update_roots(
+		src_tree_id: T::TreeId,
+		dest_tree_id: T::TreeId,
+		src_root_index: T::RootIndex,
+		dest_root_index: T::RootIndex,
+		src_deposit_root: T::Element,
+		src_withdraw_root: T::Element,
+		dest_deposit_root: T::Element,
+		dest_withdraw_root: T::Element,
+	) {
+		// Update roots
+		CachedDepositRoots::<T, I>::insert(src_tree_id, src_root_index, src_deposit_root);
+		CachedWithdrawRoots::<T, I>::insert(src_tree_id, src_root_index, src_withdraw_root);
+		CachedDepositRoots::<T, I>::insert(dest_tree_id, dest_root_index, dest_deposit_root);
+		CachedWithdrawRoots::<T, I>::insert(dest_tree_id, dest_root_index, dest_withdraw_root);
+	}
+
+	/// Claim AP tokens and update rewards VAnchor
+	pub fn claim_ap(
+		src_resource_id: ResourceId,
+		dest_resource_id: ResourceId,
+		recipient: T::AccountId,
+		amount: BalanceOf<T, I>,
+		merkle_root: T::Element,
+		latest_leaf_index: T::LeafIndex,
+		proof_data: ProofData<T::Element>,
+		src_deposit_root: T::Element,
+		src_withdraw_root: T::Element,
+		dest_deposit_root: T::Element,
+		dest_withdraw_root: T::Element,
+	) -> DispatchResultWithPostInfo {
+		// Transfer AP tokens to recipient
+		<T as Config<I>>::Currency::transfer(
+			T::AnonymityPointsAssetId::get(),
+			&Self::account_id(),
+			&recipient,
+			amount,
+		)?;
+
+		let source_tree_id: T::TreeId = match src_resource_id.target_system() {
+			TargetSystem::Substrate(system) => system.tree_id.into(),
+			_ => {
+				ensure!(false, Error::<T, I>::InvalidResourceId);
+				T::TreeId::default()
+			},
+		};
+
+		let dest_tree_id: T::TreeId = match dest_resource_id.target_system() {
+			TargetSystem::Substrate(system) => system.tree_id.into(),
+			_ => {
+				ensure!(false, Error::<T, I>::InvalidResourceId);
+				T::TreeId::default()
+			},
+		};
+
+		let deposit_root_index = Self::next_deposit_root_index();
+		let withdraw_root_index = Self::next_withdraw_root_index();
+
+		NextDepositRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+		NextWithdrawRootIndex::<T, I>::mutate(|i| {
+			*i = i.saturating_add(One::one()) % T::DepositRootHistorySize::get()
+		});
+
+		// Update roots - TODO: modify
+		Self::update_roots(
+			source_tree_id,
+			dest_tree_id,
+			deposit_root_index,
+			withdraw_root_index,
+			src_deposit_root,
+			src_withdraw_root,
+			dest_deposit_root,
+			dest_withdraw_root,
+		);
+
+		// Update rewards VAnchor
+		let vanchor =
+			Self::update_vanchor(source_tree_id, merkle_root, src_resource_id, latest_leaf_index)?;
+
+		// Handle proof verification
+		Self::handle_proof_verification(&proof_data)?;
```

Comment:
Shouldn't this line be at the beginning of the function? We have to check the proof first, before doing anything else.